reposerver/repository/repository.go

```diff
@@ -1253,6 +1362,259 @@ func getRepoCredential(repoCredentials []*v1alpha1.RepoCreds, repoURL string) *v
+	srcFullPath, err := secpath.SecureJoin(srcRoot, srcPath)
+	if err != nil {
+		logCtx.Errorf("SecureJoin(%s, %s) failed: %v", srcRoot, srcPath, err)
+		return err
+	}
+	srcStat, err := os.Lstat(srcFullPath)
+	logCtx = logCtx.WithFields(log.Fields{
+		"srcFullPath": srcFullPath,
+		"srcStat": srcStat,
+	})
+	if err != nil {
+		logCtx.Debugf("copyFilesystemSubtree: lstat failed: %v", err)
+		return err
+	}
+
+	// if the source is a symlink, we bail out, because it will certainly be out of bound
+	if srcStat.Mode()&fs.ModeSymlink >0 {
+		logCtx.Debugf("copyFilesystemSubtree: lstat failed: %v", err)
+		return fmt.Errorf("Copy source is a symlink: %s", srcFullPath)
+	}
+
+	// Check whether the source path is within the source root
+	if err := argopath.CheckPathOutOfBound(srcRoot, "", srcPath); err != nil {
+		logCtx.Debugf("copyFilesystemSubtree: argopath.CheckPathOutOfBound failed: %v", err)
+		return err
+	}
+
+	// if we are in failOnOutOfBoundsSymlink mode, check for any out of bound symlinks
+	if failOnOutOfBoundsSymlink {
+		if err := argopath.CheckOutOfBoundsSymlinks(srcFullPath); err != nil {
+			logCtx.Debugf("copyFilesystemSubtree: srcFullPath contains OOB symlinks: %v", err)
+			return err
+		}
+	}
+
+	// Checking whether the destination exists
+	dstFullPath, err := secpath.SecureJoin(dstRoot, dstPath)
+	//	dstFullPath := filepath.Join(dstRoot, dstPath)
+	if err != nil {
+		logCtx.Errorf("SecJoin(%s, %s) failed: %v", dstRoot, dstPath, err)
+		return err
+	}
+
+	dstStat, err := os.Lstat(dstFullPath)
+	dstExists := false
+	logCtx = logCtx.WithField("dstFullPath", dstFullPath)
+
+	// check destination to be within bounds:
+	if err == nil {
+		dstExists = true
+		// if the target exists check it
+		if err := argopath.CheckPathOutOfBound(dstRoot, "", dstPath); err != nil {
+			logCtx.Debugf("copyFilesystemSubtree: argopath.CheckPathOutOfBound failed: %v", err)
+			return err
+		}
+	} else if errors.Is(err, os.ErrNotExist) {
+		// if the destination doesn't yet exist, check its parent
+		if err := argopath.CheckPathOutOfBound(dstRoot, "", filepath.Dir(dstPath)); err != nil {
+			logCtx.WithField("filepath.Dir(dstPath)", filepath.Dir(dstPath)).Debugf("copyFilesystemSubtree: argopath.CheckPathOutOfBound failed: %v", err)
+			return err
+		}
+	} else {
+		return fmt.Errorf("Error while checking destination path: %v", err)
+	}
+
+	logCtx.WithFields(log.Fields{
+		"dstStat": dstStat,
+		"dstExists": dstExists,
+	}).Debug("copyFilesystemSubtree: checks passed, starting copy logic")
+
+	// if our source is just a single file
+	if srcStat.Mode().IsRegular() {
+		// in case the source is a regular file, we don't have to recurse
+		// Also, if the dst already exists, it's an error
+
+		dstFile := dstFullPath
+		// if the destination exists, it has to be a directory
+		if dstExists {
+			// if it's already there, it has to be a directory
+			if !dstStat.Mode().IsDir() {
+				return fmt.Errorf("Copy destination exists, and it's not a directory")
+			}
```

Comment:
I don't think we can have a file and a directory with same name at the same destination. We should return an error here irrespective of whether it is a directory or not.

Reply:
I beg to differ.

If the destination exists _and_ it is a directory, then there's no error, because in this codepath the source is a file, and the resulting effective destination path will be the source filename under this very directory. This is a valid usecase.

If the destination is a file, we would overwrite it, which might be a bad idea.
If it's a special device, there would be some other kind of error.
If it's a symlink, then whether there would be an error, and what kind it would be, depends on the symlink's target.

This check is to preemptive avoid later errors and ensure that the codepath is working within its intended circumstances.

If this is removed, there'll be various errors here and there, and might break some functionality, or alter them differently.

Reply:
Makes sense. Thanks